File: frontend/src/services/template_variable_resolver.ts
/* file: frontend/src/services/template_variable_resolver.ts
description: Resolver fuer konsistente Template Syntax im Frontend.
history:
- 2026-05-01: Erste Version fuer input, output, node, global und workflow Zugriff erstellt. author Marcus Schlieper
author Marcus Schlieper
*/

export type TRecord = Record<string, unknown>;

export type TTemplateResolverContext = {
  o_current_node?: TRecord;
  o_nodes_by_id?: Record<string, TRecord>;
  o_global_values?: Record<string, unknown>;
  o_workflow?: Record<string, unknown>;
};

function get_safe_record(o_value: unknown): TRecord {
  return o_value && typeof o_value === "object" ? (o_value as TRecord) : {};
}

function parse_path_segments(s_path: string): string[] {
  if (typeof s_path !== "string") {
    return [];
  }

  const a_result: string[] = [];
  const a_parts = s_path.split(".");

  for (const s_part of a_parts) {
    const s_trimmed = s_part.trim();
    if (s_trimmed === "") {
      continue;
    }

    const o_matches = s_trimmed.match(/[^[$]+|$(\d+)$/g);
    if (!o_matches) {
      a_result.push(s_trimmed);
      continue;
    }

    for (const s_match of o_matches) {
      const s_clean = s_match.replace(/^$/, "").replace(/$$/, "").trim();
      if (s_clean !== "") {
        a_result.push(s_clean);
      }
    }
  }

  return a_result;
}

function get_nested_value(o_value: unknown, s_path: string): unknown {
  if (typeof s_path !== "string" || s_path.trim() === "") {
    return o_value;
  }

  let o_current: unknown = o_value;
  const a_segments = parse_path_segments(s_path);

  for (const s_segment of a_segments) {
    if (o_current === null || typeof o_current === "undefined") {
      return undefined;
    }

    if (Array.isArray(o_current)) {
      const i_index = Number.parseInt(s_segment, 10);
      if (!Number.isInteger(i_index) || i_index < 0 || i_index >= o_current.length) {
        return undefined;
      }
      o_current = o_current[i_index];
      continue;
    }

    if (typeof o_current !== "object") {
      return undefined;
    }

    const o_record = o_current as TRecord;
    if (!(s_segment in o_record)) {
      return undefined;
    }

    o_current = o_record[s_segment];
  }

  return o_current;
}

function stringify_safe(o_value: unknown): string {
  if (o_value === null || typeof o_value === "undefined") {
    return "";
  }

  if (
    typeof o_value === "string" ||
    typeof o_value === "number" ||
    typeof o_value === "boolean" ||
    typeof o_value === "bigint"
  ) {
    return String(o_value);
  }

  try {
    return JSON.stringify(o_value);
  } catch {
    return "";
  }
}

function get_current_node_input_map(o_current_node: TRecord): TRecord {
  const o_data = get_safe_record(o_current_node.data);
  const o_named_inputs = get_safe_record(o_data.input_values);
  if (Object.keys(o_named_inputs).length > 0) {
    return o_named_inputs;
  }

  const o_result = get_safe_record(o_data.result || o_data.o_result || o_data.runtime_result);
  return get_safe_record(o_result.inputs);
}

function get_current_node_output_map(o_current_node: TRecord): TRecord {
  const o_data = get_safe_record(o_current_node.data);
  const o_output_values = get_safe_record(o_data.output_values);
  if (Object.keys(o_output_values).length > 0) {
    return o_output_values;
  }

  const o_result = get_safe_record(o_data.result || o_data.o_result || o_data.runtime_result);
  const o_output = get_safe_record(o_result.output);
  return {
    output_main: o_output,
  };
}

function get_node_runtime_view(o_node: TRecord): TRecord {
  const o_data = get_safe_record(o_node.data);
  const o_result = get_safe_record(o_data.result || o_data.o_result || o_data.runtime_result);

  return {
    input: get_safe_record(o_data.input_values),
    output: get_safe_record(o_data.output_values),
    result: o_result,
    status: o_data.s_runtime_status,
  };
}

export function resolve_template_variable(
  s_expression: string,
  o_context: TTemplateResolverContext
): unknown {
  const s_safe_expression = typeof s_expression === "string" ? s_expression.trim() : "";
  if (s_safe_expression === "") {
    return undefined;
  }

  const i_separator = s_safe_expression.indexOf(":");
  if (i_separator <= 0) {
    return undefined;
  }

  const s_scope = s_safe_expression.slice(0, i_separator).trim();
  const s_body = s_safe_expression.slice(i_separator + 1).trim();

  if (s_scope === "input") {
    const o_current_node = get_safe_record(o_context.o_current_node);
    const o_input_map = get_current_node_input_map(o_current_node);
    return get_nested_value(o_input_map, s_body);
  }

  if (s_scope === "output") {
    const o_current_node = get_safe_record(o_context.o_current_node);
    const o_output_map = get_current_node_output_map(o_current_node);
    return get_nested_value(o_output_map, s_body);
  }

  if (s_scope === "global") {
    return get_nested_value(get_safe_record(o_context.o_global_values), s_body);
  }

  if (s_scope === "workflow") {
    return get_nested_value(get_safe_record(o_context.o_workflow), s_body);
  }

  if (s_scope === "node") {
    const i_first_dot = s_body.indexOf(".");
    if (i_first_dot <= 0) {
      return undefined;
    }

    const s_node_id = s_body.slice(0, i_first_dot).trim();
    const s_node_path = s_body.slice(i_first_dot + 1).trim();
    const o_nodes_by_id = o_context.o_nodes_by_id || {};
    const o_node = get_safe_record(o_nodes_by_id[s_node_id]);
    const o_runtime_view = get_node_runtime_view(o_node);

    return get_nested_value(o_runtime_view, s_node_path);
  }

  return undefined;
}

export function resolve_template_string(
  s_template: string,
  o_context: TTemplateResolverContext
): string {
  if (typeof s_template !== "string" || s_template.trim() === "") {
    return "";
  }

  return s_template.replace(/\{\{([^{}]+)\}\}/g, (_s_match: string, s_expression: string) => {
    const o_resolved = resolve_template_variable(s_expression, o_context);
    return stringify_safe(o_resolved);
  });
}

File: backend/services/template_variable_resolver.py
# file: backend/services/template_variable_resolver.py
# description: Sichere Aufloesung der konsistenten Template Syntax im Backend.
# history:
# - 2026-05-01: Erste Version fuer input, output, node, global und workflow Zugriff erstellt. author Marcus Schlieper
# author Marcus Schlieper

import re
from typing import Any, Dict, List


def _get_safe_dict(o_value: Any) -> Dict[str, Any]:
    return o_value if isinstance(o_value, dict) else {}


def _parse_path_segments(s_path: str) -> List[str]:
    if not isinstance(s_path, str):
        return []

    a_result: List[str] = []
    for s_part in s_path.split("."):
        s_part = s_part.strip()
        if s_part == "":
            continue

        a_matches = re.findall(r"[^.$$]+|$(\d+)$", s_part)
        if not a_matches:
            a_result.append(s_part)
            continue

        for o_match in re.finditer(r"[^$$]+|$(\d+)$", s_part):
            s_token = o_match.group(0)
            if s_token.startswith("[") and s_token.endswith("]"):
                s_token = s_token[1:-1].strip()
            else:
                s_token = s_token.strip()

            if s_token != "":
                a_result.append(s_token)

    return a_result


def _get_nested_value(o_value: Any, s_path: str) -> Any:
    if not isinstance(s_path, str) or s_path.strip() == "":
        return o_value

    o_current = o_value
    a_segments = _parse_path_segments(s_path)

    for s_segment in a_segments:
        if o_current is None:
            return None

        if isinstance(o_current, list):
            if not s_segment.isdigit():
                return None
            i_index = int(s_segment)
            if i_index < 0 or i_index >= len(o_current):
                return None
            o_current = o_current[i_index]
            continue

        if not isinstance(o_current, dict):
            return None

        if s_segment not in o_current:
            return None

        o_current = o_current[s_segment]

    return o_current


def _stringify_safe(o_value: Any) -> str:
    if o_value is None:
        return ""

    if isinstance(o_value, (str, int, float, bool)):
        return str(o_value)

    try:
        import json
        return json.dumps(o_value, ensure_ascii=True)
    except Exception:
        return ""


def resolve_template_variable(
    s_expression: str,
    o_current_node_context: Dict[str, Any],
    d_results_by_node_id: Dict[str, Dict[str, Any]],
    d_global_values: Dict[str, Any],
    d_workflow_values: Dict[str, Any],
) -> Any:
    s_safe_expression = str(s_expression).strip()
    if s_safe_expression == "" or ":" not in s_safe_expression:
        return None

    s_scope, s_body = s_safe_expression.split(":", 1)
    s_scope = s_scope.strip()
    s_body = s_body.strip()

    if s_scope == "input":
        d_named_inputs = _get_safe_dict(o_current_node_context.get("named_inputs", {}))
        return _get_nested_value(d_named_inputs, s_body)

    if s_scope == "output":
        d_current_result = _get_safe_dict(o_current_node_context.get("current_result", {}))
        d_output_meta = _get_safe_dict(d_current_result.get("output_meta", {}))
        d_node_outputs = _get_safe_dict(d_output_meta.get("node_outputs", {}))
        return _get_nested_value(d_node_outputs, s_body)

    if s_scope == "global":
        return _get_nested_value(_get_safe_dict(d_global_values), s_body)

    if s_scope == "workflow":
        return _get_nested_value(_get_safe_dict(d_workflow_values), s_body)

    if s_scope == "node":
        if "." not in s_body:
            return None

        s_node_id, s_node_path = s_body.split(".", 1)
        s_node_id = s_node_id.strip()
        s_node_path = s_node_path.strip()

        d_node_wrapper = _get_safe_dict(d_results_by_node_id.get(s_node_id, {}))
        d_node_result = _get_safe_dict(d_node_wrapper.get("result", {}))
        d_node_inputs = _get_safe_dict(d_node_wrapper.get("inputs", {}))
        d_output_meta = _get_safe_dict(d_node_result.get("output_meta", {}))
        d_node_outputs = _get_safe_dict(d_output_meta.get("node_outputs", {}))

        d_runtime_view = {
            "input": _get_safe_dict(d_node_inputs.get("named_inputs", {})),
            "output": d_node_outputs,
            "result": d_node_result,
            "status": d_node_wrapper.get("status"),
        }
        return _get_nested_value(d_runtime_view, s_node_path)

    return None


def resolve_template_text(
    s_template: str,
    o_current_node_context: Dict[str, Any],
    d_results_by_node_id: Dict[str, Dict[str, Any]],
    d_global_values: Dict[str, Any],
    d_workflow_values: Dict[str, Any],
) -> str:
    s_safe_template = str(s_template or "")

    def _replace(o_match: re.Match[str]) -> str:
        s_expression = o_match.group(1)
        o_value = resolve_template_variable(
            s_expression=s_expression,
            o_current_node_context=o_current_node_context,
            d_results_by_node_id=d_results_by_node_id,
            d_global_values=d_global_values,
            d_workflow_values=d_workflow_values,
        )
        return _stringify_safe(o_value)

    return re.sub(r"\{\{([^{}]+)\}\}", _replace, s_safe_template)
